src/runtime/virtcontainers/fs_share.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2022 Apple Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"context"
+
+	"github.com/kata-containers/kata-containers/src/runtime/virtcontainers/pkg/agent/protocols/grpc"
+)
+
+// fsShareTracingTags defines tags for the trace span
+var fsShareTracingTags = map[string]string{
+	"source":    "runtime",
+	"package":   "virtcontainers",
+	"subsystem": "fs_share",
+}
+
+// SharedFile represents the outcome of a host filesystem sharing
+// operation.
+type SharedFile struct {
+	storage   *grpc.Storage
+	guestPath string
+}
+
+type FilesystemSharer interface {
+	// Prepare will set the host filesystem up, making it ready
+	// to share container files and directories with the guest.
+	// Implementation of this method must be idempotent and be
+	// ready to potentially be called several times in a row,
+	// without symetric calls to Cleanup in between.
+	Prepare(context.Context) error
+
+	// Cleanup cleans the host filesystem up from the initial
+	// setup created by Prepare.
+	// Implementation of this method must be idempotent and be
+	// ready to potentially be called several times in a row,
+	// without symetric calls to Prepare in between.
+	Cleanup(context.Context) error
+
+	// ShareFile shares a file (a regular file or a directory)
+	// from the host filesystem with a container running in the
+	// guest. The host file to be shared is described by the
+	// Mount argument.
+	// The returned SharedFile pointer describes how the file
+	// should be shared between the host and the guest. If it
+	// is nil, then the shared filed described by the Mount
+	// argument will be ignored by the guest, i.e. it will NOT
+	// be shared with the guest.
```

Comment:
Two questions:
1. Is it OK to call it multiple times with different `Mount`s, and if so, what is the result
2. Since this returns a `SharedFile`, this means the implementations decide the strategy on what can be shared. Doesn't that imply that the description should be "i.e. it _can_ NOT been shared with the guest" instead of "it _will_ not"?

Reply:
> Two questions:
> 
>     1. Is it OK to call it multiple times with different `Mount`s, and if so, what is the result

It's more than ok, it's what implementation would do. The description states that this method is for sharing one file (regular or directory), and that the file is described through the `Mount` argument:

```
        // ShareFile shares a file (a regular file or a directory)
        // from the host filesystem with a container running in the
        // guest. The host file to be shared is described by the
        // Mount argument.
```

I guess this is not clear enough, I extended it to be more explicit.

>     2. Since this returns a `SharedFile`, this means the implementations decide the strategy on what can be shared. Doesn't that imply that the description should be "i.e. it _can_ NOT been shared with the guest" instead of "it _will_ not"?
 
Both. It can not and thus will not. I tried to make that clearer.